vendor_skins/Evercast/plugins/websocket-client/EvercastWebsocketClientImpl.cpp

```diff
@@ -141,10 +143,10 @@ bool EvercastWebsocketClientImpl::connect(
         });
 
         // --- Close handler
-        client.set_close_handler(std::bind(&EvercastWebsocketClientImpl::handleDisconnect, this, _1, listener, "set_close_handler"));
+        client.set_close_handler(std::bind(&EvercastWebsocketClientImpl::handleDisconnect, this, _1, listener));
 
         // --- Failure handler
-        client.set_fail_handler(std::bind(&EvercastWebsocketClientImpl::handleDisconnect, this, _1, listener, "set_fail_handler"));
+        client.set_fail_handler(std::bind(&EvercastWebsocketClientImpl::handleFail, this, _1, listener));
```

Comment:
Why is this one different?

Reply:
Sorry, missed this when you originally commented.  The handlers are different now because there are cases when you don't want the fail handler to disconnect, specifically because a fail notification can happen as the result of an error during a message receipt - attempting to disconnect while in the middle of processing a message causes threadlock, competing for access to the mutex used to synchronize access to the websocket.  This is because disconnect tries to send a message of its own to the server, a close message specifically.  So we have a softer handler here, which logs the error on failure, but doesn't disconnect.  Most failures should result in disconnection (and indeed many arise as the result of an unplanned disconnect), but this is done separately by infrastructure; this handler is not responsible for it.

It is important to keep this distinct from cleaning up the local resources of course.  That is very much a separate task from disconnection, and since the two are interrelated, must be handled in conjunction with the disconnection of the socket.  Errors in this process are what led to earlier problems with stability; a second (third, fourth, etc.) pair of eyes would therefore be very welcome when it comes to this code, particularly around different sorts of failures.

Reply:
(Addendum: I looked up the code to get a proper picture for you.  The onLoggedError callback in fact cleans up the resources itself, which is good.  It just does this in a non-blocking fashion, which allows this to clean up its resources smoothly, rather than hanging.)